pkg/webhook/controlplane/genericmutator/mutator.go

```diff
@@ -102,6 +108,39 @@ func (m *mutator) InjectClient(client client.Client) error {
 	return nil
 }
 
+type ensurerContext struct {
+	client  client.Client
+	object  metav1.Object
+	cluster *extensionscontroller.Cluster
+}
+
+// NewEnsurerContext creates a ensurer context object.
+func NewEnsurerContext(client client.Client, object metav1.Object) EnsurerContext {
+	return &ensurerContext{
+		client: client,
+		object: object,
+	}
+}
+
+// NewInternalEnsurerContext creates a ensurer context object.
+func NewInternalEnsurerContext(cluster *extensionscontroller.Cluster) EnsurerContext {
```

Comment:
Maybe we can use only `NewEnsurerContext` for everything and pass `nil` for un-necessary values. Instead of calling `NewInternalContext` to just get the cluster just call `NewEnsurerContext` with a nil value for the client but have always one instantiation function ?

Reply:
Not sure why this is a problem. The internal one is for our tests only as we need to mock a cluster object into the the whole thing in order to run some of the tests.

Reply:
To me it is a little confusing to have exposed Initializers for the same object just to fill in different fields for a test. My proposal was to have one initializer, later in the test you would do something like:  
```go 
dummyContext = genericmutator.NewEnsurerContext(nil, nil, cluster)
``` 
Also Just my opinion :)

Reply:
sounds better to me, we had a similar approach here https://github.com/gardener/gardener-resource-manager/blob/7ceeb30219932467612119392067cacabb9aaf6f/pkg/manager/managedresources.go#L33. 

`NewEnsurerContext().WithCluster(..).WithClient(...)`

Reply:
A function signature should always be oriented on the intended semantic. For the API function this semantic explicitly does NOT allow to set a cluster object, therefor the signature has no such argument. Only for the test purposes we need to create fake contexts beside the official API usage (White box testing). Therfore we added an additional function for this purpose not intended be part of the Official API.

Reply:
Sure, but here if I am an external user to the API and I find this function, I would not know that it is intended only for tests, also the name 'Internal' contradicts with the function scope. I guess for such use-case an `internal` package can be used to prevent exposing the function to other external actors. Or simply re-use the same initializer. 

That said, this is just an opinion on my side, feel free to ignore if you don't think it adds value to your PR :)

Reply:
I'm okay getting it in as it is, there are also other PRs waiting on this one. @zanetworker also ok?

Reply:
Also, can't we just refactor the test later on? There seems to be different opinions on how to do that. Maybe we should discuss first?

Reply:
Sure, please go ahead.